prov/util/src/util_mem_monitor.c

```diff
@@ -140,6 +144,8 @@ int ofi_monitor_add_cache(struct ofi_mem_monitor *monitor,
 			ret = ofi_uffd_init();
 		else if (monitor == memhooks_monitor)
 			ret = ofi_memhooks_init();
+		else if (monitor == cuda_monitor)
+			ret = FI_SUCCESS;
```

Comment:
I would modify the cache to support 2 (or more) monitors, up to 1 per fi_hmem_iface.  At a minimum, I would allow the host memory monitor, plus one device monitor.  Though it's likely barely more work or overhead to handle all monitors at once.  The monitor would be selected based on the ofi_mr_entry hmem_iface value for each region.

We need a single cache for each virtual/device address range that we're dealing with.  If we have a unified virtual address range, then those addresses should all be stored in the same cache.  If we're dealing with GPU specific addresses (where the value returned could overlap with a value returned through a host memory call), then those need to go into a separate cache.

If we can defer dealing with multiple caches until a later patch set, that would be preferable to me.  But I think that information needs to come down from the application through the API, with the providers selecting the correct cache.

Reply:
I was starting to look into this change. Would you like the "cache to support 2 (or more) monitors, up to 1 per fi_hmem_iface" change as a part of this patch set or a future PR?

Reply:
I question whether we can get a working solution without having the cache support multiple monitors, so I think it needs to be part of an initial set of changes.  That should also minimize the changes needed to the providers.

I'm okay with unified virtual addressing being required in order to use the cache.  We can document that restriction.  I think this is what works for most applications though.

Trying to support non-unified addressing requires a bit more work to make sure we can get that correct, and requires application changes.

Reply:
> I question whether we can get a working solution without having the cache support multiple monitors, so I think it needs to be part of an initial set of changes. That should also minimize the changes needed to the providers.

I'll add another commit to this PR for the implementation.

> I'm okay with unified virtual addressing being required in order to use the cache. We can document that restriction. I think this is what works for most applications though.

I can document this restriction.

Reply:
We can separate the changes to util_mem_monitor.c.  Those changes will need to have the multiple monitor support.

Reply:
I'll put these util_mem_monitor.c changes into their own commit.

Reply:
To be clear, I don't think we want the above changes.  We need to update the cache to support multiple monitors, which I think will result in a fair number of improvements to the code.

This is just a guess at this point, but for example, I think we may end up changing ofi_mem_monitor to something like:

```
struct ofi_mem_monitor {
    blah blah existing fields... plus new valid() callback

    int (*init)(struct ofi_mem_monitor *monitor);
    void (*cleanup)(struct ofi_mem_monitor *monitor);
};
```

This would allow the cache to call init/cleanup in a loop for all active monitors.  The init() / cleanup() calls would need to be statically set.  E.g.

```
static struct ofi_uffd uffd = {
    .init = ofi_uffd_init,
    .cleanup = ofi_uffd_cleanup
};
```

The ofi_mr_cache would be updated to:

```
struct ofi_mr_cache {
    struct util_domain *domain;
    struct ofi_mem_monitor *monitor[OFI_HMEM_MAX];
    ....
};
```

The mr_cache->monitor[FI_HMEM_SYSTEM] would be set to either the uffd_monitor or memhook_monitor similar to how it is set today.  mr_cache->monitor[FI_HMEM_CUDA] would point to the only cuda option we currently have.

I haven't thought through the details, so it's possible I'm overlooking something that would make this approach impractical, such as handling a disabled monitor.

Reply:
This is similar to the approach I took in the _core/mr_cache: Support multiple monitors per cache_ commit. I'll look at adding in the init and cleanup ops to ofi_mem_monitor.

Reply:
I'm going to make a change to the existing code that I think will help make the memory monitor a little clearer (because I'm confusing myself).  As an example, the gist is to do the following:

ofi_memhooks_init -> rename -> ofi_memhooks_start
ofi_memhooks_cleanup -> rename -> ofi_memhooks_stop

Create new ofi_memhooks_init call with code from ofi_monitor_init.

A similar change would be made for uffd.  I don't think we need cleanup routines, but we can add those if needed.  This should help with the confusion of having 2 init type locations for the memory monitors.

It should only take me a few hours to make this change and open a PR.

Reply:
Sounds good.